File: franka_simulink/blocks/franka_robot_model_mass_sfunction.cpp
/*
 * franka_robot_model_mass_sfunction.cpp - C++ Level-2 S-Function
 *
 * Computes the 7x7 mass matrix M(q) using libfranka's Model class.
 *
 * Inputs:
 *   0. robot_ip   (16x1 uint8)  - Robot IP as ASCII chars
 *   1. q          (7x1 double)  - Joint positions [rad]
 *   2. I_total    (3x3 double)  - Total load inertia [kg*m^2]
 *   3. m_total    (1x1 double)  - Total load mass [kg]
 *   4. F_x_Ctotal (3x1 double)  - Total load CoM in flange frame [m]
 *
 * Outputs:
 *   0. mass       (7x7 double)  - Mass matrix M(q) [kg*m^2]
 *
 * Copyright (c) 2025 Franka Robotics GmbH
 */

#define S_FUNCTION_NAME  franka_robot_model_mass_sfunction
#define S_FUNCTION_LEVEL 2

#include "simstruc.h"

/* Parameters - none */
#define NUM_PARAMS  0

/* Input port indices */
#define IN_ROBOT_IP     0   /* Robot IP (16x1 uint8) */
#define IN_Q            1   /* Joint positions (7x1) */
#define IN_I_TOTAL      2   /* Load inertia (3x3) */
#define IN_M_TOTAL      3   /* Load mass (1x1) */
#define IN_F_X_CTOTAL   4   /* Load CoM (3x1) */
#define NUM_INPUTS      5

/* Output port indices */
#define OUT_MASS        0   /* Mass matrix (7x7) */
#define NUM_OUTPUTS     1

/* ========================================================================
 * mdlInitializeSizes - Initialize block sizes
 * ======================================================================== */
static void mdlInitializeSizes(SimStruct *S)
{
    ssSetNumSFcnParams(S, NUM_PARAMS);
    
#if defined(MATLAB_MEX_FILE)
    if (ssGetNumSFcnParams(S) != ssGetSFcnParamsCount(S)) {
        return;
    }
#endif

    ssSetNumContStates(S, 0);
    ssSetNumDiscStates(S, 0);
    
    /* INPUT PORTS */
    if (!ssSetNumInputPorts(S, NUM_INPUTS)) return;
    
    /* Port 0: robot_ip (16 ASCII chars) */
    ssSetInputPortWidth(S, IN_ROBOT_IP, 16);
    ssSetInputPortDataType(S, IN_ROBOT_IP, SS_UINT8);
    ssSetInputPortDirectFeedThrough(S, IN_ROBOT_IP, 1);
    ssSetInputPortRequiredContiguous(S, IN_ROBOT_IP, 1);
    
    /* Port 1: q - Joint positions (7x1) */
    ssSetInputPortWidth(S, IN_Q, 7);
    ssSetInputPortDataType(S, IN_Q, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_Q, 1);
    ssSetInputPortRequiredContiguous(S, IN_Q, 1);
    
    /* Port 2: I_total - Load inertia (3x3 = 9 elements) */
    ssSetInputPortMatrixDimensions(S, IN_I_TOTAL, 3, 3);
    ssSetInputPortDataType(S, IN_I_TOTAL, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_I_TOTAL, 1);
    ssSetInputPortRequiredContiguous(S, IN_I_TOTAL, 1);
    
    /* Port 3: m_total - Load mass (1x1) */
    ssSetInputPortWidth(S, IN_M_TOTAL, 1);
    ssSetInputPortDataType(S, IN_M_TOTAL, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_M_TOTAL, 1);
    ssSetInputPortRequiredContiguous(S, IN_M_TOTAL, 1);
    
    /* Port 4: F_x_Ctotal - Load CoM (3x1) */
    ssSetInputPortWidth(S, IN_F_X_CTOTAL, 3);
    ssSetInputPortDataType(S, IN_F_X_CTOTAL, SS_DOUBLE);
    ssSetInputPortDirectFeedThrough(S, IN_F_X_CTOTAL, 1);
    ssSetInputPortRequiredContiguous(S, IN_F_X_CTOTAL, 1);
    
    /* OUTPUT PORTS */
    if (!ssSetNumOutputPorts(S, NUM_OUTPUTS)) return;
    
    /* Port 0: mass matrix (7x7 = 49 elements) */
    ssSetOutputPortMatrixDimensions(S, OUT_MASS, 7, 7);
    ssSetOutputPortDataType(S, OUT_MASS, SS_DOUBLE);
    
    /* SAMPLE TIME */
    ssSetNumSampleTimes(S, 1);
    
    /* WORK VECTORS */
    ssSetNumRWork(S, 0);
    ssSetNumIWork(S, 0);
    ssSetNumPWork(S, 0);
    ssSetNumModes(S, 0);
    ssSetNumNonsampledZCs(S, 0);
    ssSetNumDWork(S, 0);
    
    /* Options */
    ssSetSimStateCompliance(S, USE_DEFAULT_SIM_STATE);
    ssSetOptions(S, SS_OPTION_CALL_TERMINATE_ON_EXIT);
}

/* ========================================================================
 * mdlInitializeSampleTimes - Initialize sample times
 * ======================================================================== */
static void mdlInitializeSampleTimes(SimStruct *S)
{
    ssSetSampleTime(S, 0, INHERITED_SAMPLE_TIME);
    ssSetOffsetTime(S, 0, 0.0);
}

/* ========================================================================
 * mdlOutputs - Compute outputs (simulation only)
 * ======================================================================== */
static void mdlOutputs(SimStruct *S, int_T tid)
{
    /* For simulation only - actual implementation is in TLC */
    real_T *mass = ssGetOutputPortRealSignal(S, OUT_MASS);
    
    /* Zero output for simulation */
    for (int i = 0; i < 49; i++) {
        mass[i] = 0.0;
    }
    
    UNUSED_ARG(tid);
}

/* ========================================================================
 * mdlTerminate - Cleanup
 * ======================================================================== */
static void mdlTerminate(SimStruct *S)
{
    UNUSED_ARG(S);
}

/* ========================================================================
 * Required S-function trailer
 * ======================================================================== */
#ifdef MATLAB_MEX_FILE
#include "simulink.c"
#else
#include "cg_sfun.h"
#endif

File: franka_simulink/blocks/franka_robot_model_mass_sfunction.m
function franka_robot_model_mass_sfunction(block)
%FRANKA_ROBOT_MODEL_MASS Compute mass matrix M(q) using libfranka Model
%
%  This block computes the 7x7 mass matrix from explicit inputs using
%  the robot's kinematic/dynamic model. It integrates with FrankaRobotManager
%  to share the robot connection with other blocks.
%
%  Parameters:
%    robot_ip - IP address of the Franka robot (for model lookup)
%
%  Inputs:
%    q          - Joint positions [rad] (7x1)
%    I_total    - Total load inertia [kg*m^2] (3x3 column-major as 9x1)
%    m_total    - Total load mass [kg] (1x1)
%    F_x_Ctotal - Total load CoM in flange frame [m] (3x1)
%
%  Outputs:
%    mass - Mass matrix M(q) [kg*m^2] (7x7)
%
%  Usage Notes:
%    - For current robot state model data, use the model_data output from
%      franka_robot block instead (FrankaModelDataBus.mass)
%    - This block is for computing model data from arbitrary configurations
%    - Works both standalone AND under function-called subsystems
%    - Attempts to connect to robot on first execution if not already connected
%
%  Copyright (c) 2025 Franka Robotics GmbH
%  This file is subject to the terms and conditions defined in the file
%  'LICENSE', which is part of this package

setup(block);

function setup(block)

    % Register parameters: robot_ip only
    block.NumDialogPrms     = 1;
    block.DialogPrmsTunable = {'Nontunable'};
    
    % Register number of ports
    % Inputs: q(7), I_total(9), m_total(1), F_x_Ctotal(3)
    block.NumInputPorts  = 4;
    block.NumOutputPorts = 1;

    % Setup port properties to be inherited or dynamic
    block.SetPreCompInpPortInfoToDynamic;
    block.SetPreCompOutPortInfoToDynamic;

    % Input 1: q - Joint positions (7x1)
    block.InputPort(1).Dimensions  = 7;
    block.InputPort(1).DatatypeID  = 0;  % double
    block.InputPort(1).Complexity  = 'Real';
    block.InputPort(1).DirectFeedthrough = true;

    % Input 2: I_total - Total load inertia (9x1, 3x3 column-major)
    block.InputPort(2).Dimensions  = 9;
    block.InputPort(2).DatatypeID  = 0;  % double
    block.InputPort(2).Complexity  = 'Real';
    block.InputPort(2).DirectFeedthrough = true;

    % Input 3: m_total - Total load mass (1x1)
    block.InputPort(3).Dimensions  = 1;
    block.InputPort(3).DatatypeID  = 0;  % double
    block.InputPort(3).Complexity  = 'Real';
    block.InputPort(3).DirectFeedthrough = true;

    % Input 4: F_x_Ctotal - Total load CoM (3x1)
    block.InputPort(4).Dimensions  = 3;
    block.InputPort(4).DatatypeID  = 0;  % double
    block.InputPort(4).Complexity  = 'Real';
    block.InputPort(4).DirectFeedthrough = true;

    % Output: mass matrix (7x7 column-major)
    block.OutputPort(1).Dimensions  = [7, 7];
    block.OutputPort(1).DatatypeID  = 0;  % double
    block.OutputPort(1).Complexity  = 'Real';
    block.OutputPort(1).SamplingMode = 'Sample';
    
    % Inherited sample time: works in both standalone and function-called contexts
    % - Standalone: inherits from driving source (e.g., constant blocks at model rate)
    % - Function-called subsystem: inherits from fcall trigger
    block.SampleTimes = [-1 0];

    block.SimStateCompliance = 'DefaultSimState';

    % Register methods
    block.RegBlockMethod('CheckParameters',         @CheckPrms);
    block.RegBlockMethod('ProcessParameters',       @ProcessPrms);
    block.RegBlockMethod('PostPropagationSetup',    @DoPostPropSetup);
    block.RegBlockMethod('Start',                   @Start); 
    block.RegBlockMethod('WriteRTW',                @WriteRTW);
    block.RegBlockMethod('Outputs',                 @Outputs);

function CheckPrms(block)
    % Validate robot_ip parameter
    robot_ip = block.DialogPrm(1).Data;
    if ~ischar(robot_ip) && ~isstring(robot_ip)
        error('robot_ip must be a string');
    end

function DoPostPropSetup(block)
    % No DWork needed for this simplified version

function ProcessPrms(block)
    block.AutoUpdateRuntimePrms;

function Start(block)
    % Initialization - nothing to do here
    % Connection is handled lazily by FrankaRobotManager in generated code

function Outputs(block)
    % For simulation only - actual implementation is in TLC
    % Output zeros during simulation
    block.OutputPort(1).Data = zeros(7, 7);

function WriteRTW(block)
    % Write parameters to RTW file for TLC access
    robot_ip = block.DialogPrm(1).Data;
    
    % Format robot_ip for TLC
    robot_ip = char(['''', robot_ip, '''']);
    
    block.WriteRTWParam('string', 'robot_ip', robot_ip);
